mathics/builtin/lists.py

```diff
@@ -2096,14 +2099,235 @@ def apply(self, list, sep, evaluation):
+        self._groups = []
+        self._test = test
+        self._evaluation = evaluation
+        self._name = name
+
+    def selector(self):
+        groups = self._groups
+        return lambda elem: groups
+
+    def same(self):
+        test = self._test
+        evaluation = self._evaluation
+        return _make_test_pair(test, evaluation, self._name)
+
+
+class _FastEquivalence:
+    # models an equivalence relation through SameQ. for n distinct elements (each
+    # in its own bin), we expect to make O(n) comparisons (if the hash function
+    # does not fail us by distributing items very unevenly).
+
+    # IMPORTANT NOTE ON ATOM'S HASH FUNCTIONS / this code relies on this assumption:
+    #
+    # if SameQ[a, b] == true then hash(a) == hash(b)
+    #
+    # more specifically, this code bins items based on their hash code, and only if
+    # the hash code matches, is SameQ evoked.
+    #
+    # this assumption has been checked for these types: Integer, Real, Complex,
+    # String, Rational (*), Expression, Image; new atoms need proper hash functions
+    #
+    # (*) Rational values are sympy Rationals which are always held in reduced form
+    # and thus are hashed correctly (see sympy/core/numbers.py:Rational.__eq__()).
+
+    def __init__(self):
+        self._hashes = defaultdict(list)
+
+    def selector(self):
+        hashes = self._hashes
+        return lambda elem: hashes[hash(elem)]
+
+    def same(self):
+        return lambda a, b: a.same(b)
+
+
+class _GatherBin:
+    def __init__(self, item):
+        self._items = [item]
+        self.add_to = self._items.append
+
+    def from_python(self):
+        return Expression('List', *self._items)
+
+
+class _TallyBin:
+    def __init__(self, item):
+        self._item = item
+        self._count = 1
+
+    def add_to(self, item):
+        self._count += 1
+
+    def from_python(self):
+        return Expression('List', self._item, Integer(self._count))
+
+
+class _DeleteDuplicatesBin:
+    def __init__(self, item):
+        self._item = item
+        self.add_to = lambda elem: None
+
+    def from_python(self):
+        return self._item
+
+
+class _GatherOperation(Builtin):
+    rules = {
+        '%(name)s[list_]': '%(name)s[list, SameQ]'
+    }
+
+    messages = {
+        'normal': 'Nonatomic expression expected at position `1` in `2`.',
+        'needlist': 'expecting a List as first parameter.',
+        'smtst': ("Application of the SameTest yielded `1`, which evaluates "
+                  "to `2`. The SameTest must evaluate to True or False at "
+                  "every pair of elements."),
+    }
+
+    def apply(self, list, test, evaluation):
+        '%(name)s[list_, test_]'
+
+        if list.is_atom():
+            expr = Expression(self.get_name(), list, test)
+            evaluation.message(self.get_name(), 'normal', 1, expr)
+            return
+
+        if list.get_head_name() != 'System`List':
+            evaluation.error(self.get_name(), 'needlist')
```

Comment:
`evaluation.message` acts as both a warning and an error message. It's used whenever evaluation can continue. Normally continuing just means return the expression unevaluated. In this case for example, there is no evaluation 'error', it's just that the function doesn't apply to all inputs.

`evaluation.error` on the other hand is used when there is an error with the evaluation process itself. For example, if the recursion depth gets above `$RecursionLimit` then that's an evaluation error since evaluation can't continue. We only use `evauation.error` when the calculation has to be aborted.

We try to keep things compatible with Mathematica.

Does that make sense?